src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs

```diff
@@ -183,8 +183,11 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol?
             {
                 // Limit the search space to the outermost code block that could contain references to this expr (or
                 // fall back to compilation unit for top level statements).
-                var outermostBody = invokedExpression.AncestorsAndSelf().Last(
-                    n => n is BlockSyntax or ArrowExpressionClauseSyntax or AnonymousFunctionExpressionSyntax or CompilationUnitSyntax);
+                var outermostBody = invokedExpression.AncestorsAndSelf().LastOrDefault(
+                    n => n is BlockSyntax or ArrowExpressionClauseSyntax or AnonymousFunctionExpressionSyntax or GlobalStatementSyntax);
+                if (outermostBody is null or GlobalStatementSyntax)
+                    outermostBody = syntaxTree.GetRoot(cancellationToken);
```

Comment:
the prior walk *always* hit the compilation unit.  which meant that we were always checking everything (Even unrelated code).  now we keep walking up, and only go to the compilation unit if we were in a global statement.